library/core/tests/num/int_macros.rs

```diff
@@ -338,6 +338,32 @@ macro_rules! int_module {
                 assert_eq!(MIN.checked_next_multiple_of(-3), None);
                 assert_eq!(MIN.checked_next_multiple_of(-1), Some(MIN));
             }
+
+            #[test]
+            fn test_carrying_add() {
+                assert_eq!($T::MAX.carrying_add(1, false), ($T::MIN, true));
+                assert_eq!($T::MAX.carrying_add(0, true), ($T::MIN, true));
+                assert_eq!($T::MAX.carrying_add(1, true), ($T::MIN + 1, true));
+                assert_eq!($T::MAX.carrying_add(-1, false), ($T::MAX - 1, false));
+                assert_eq!($T::MAX.carrying_add(-1, true), ($T::MAX, false)); // no intermediate overflow
+                assert_eq!($T::MIN.carrying_add(-1, false), ($T::MAX, true));
```

Comment:
The piece that's [still](https://github.com/rust-lang/rust/issues/90541#issuecomment-966895088) unclear to me: how is it helpful for something labelled `bigint_helper_methods` for both `MIN + -1` and `MAX + 1` to have the same `true` carry?  It seems to me like it's fundamentally not enough information to "allows for chaining together multiple additions" (like the doc comment says the method can be used to do).

Reply:
In the example I give in https://github.com/rust-lang/rust/issues/85532#issuecomment-967523098, to implement the `overflowing_add/_sub` operations on bit ints, you first need to perform the add operation on the unsigned low order datas first.
When `carrying_add`/`borrowing_sub` to/from the highest order data, which is signed for signed bit ints, the input `bool` comes from an unsigned overflow, but the output inform on signed overflow, which is exactly the return value necessary to implement the `overflowing_add/_sub` on bit ints.

To summarize, when implementing bit ints, the workflow is to chain the unsigned overflow from the low order datas to the higher order ones, and when performing the last operation, the result is not chained, but in fact used to indicate what happened during the operation.
**In other words, you will never chain a signed overflow, because it's not indented to be chained. It's intended to inform on the result of the overall operation.**

In my experience with the Motorola 68000, this is the intended behaviour on big ints. The unsigned overflow flag is chained across `ADD` operations to perform the addition on a bit int. During the chaining, the signed overflow flag is not used because it is not valid. It becomes valid after the last `ADD` instruction occured, because the big int implemented is signed.

I hope this explaination is clear.

Reply:
>  It seems to me like it's fundamentally not enough information to "allows for chaining together multiple additions" (like the doc comment says the method can be used to do).

I didn't understood this sentence correctly. I've changed the documentation on signed big int explaining their purpose along with an example. Sorry for the confusion, that's my bad.

The example makes the doc pretty long, so I'm not sure if it's necessary to keep it or not.

Reply:
Based on an example I made, the implementations are way different between multiplication and addition/subtraction. I think the signature of `carrying_mul` (and by extension `widening_mul` too) would be
```rs
fn carrying_mul(multiplier: i32, multiplicand: i32, carry: i32) -> (i32 /* product */, i32 /* carry */);
fn widening_mul(multiplier: i32, multiplicand: i32) -> (i32 /* product */, i32 /* carry */);
```
You can see my own implementation of `widening_mul` on a I16 big int here : https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=274a2456f969f70c844d0579c4b97faa
It is the same code for both signed and unsigned big ints.

The idea of the implementation is that multiplying two big ints together (for example I16, which is made of an u8 and an i8): `lhs * rhs`  results in computing `(lhs.high + lhs.low) * (rhs.high + rhs.low)`. When we distribute the factors, we get
```
lhs.high * rhs.high + lhs.high * rhs.low + lhs.low * rhs.high + lhs.low * rhs.low
```
What may look strange in signed contexts is the fact that the multiplicand is cast from unsigned to signed before being used, but it is mandatory to use signed semantics to obtain the correct result.

If you want to I can provide a complete breakdown on what happens when running this code on both a signed big int and an unsigned big int.